### PR TITLE
fix: MrVI MixtureSameFamily log probability fixed

### PR DESCRIPTION
**Pull request by justjhong**
Fixes https://github.com/scverse/scvi-tools/issues/3188.

This PR fixes the issue in both the `get_aggregated_posterior` function as well as the Mixture of Gaussians prior.

Tested on tutorial: integration remains the same, differential abundance results now take a much larger range as expected (see issue). To mitigate spikiness in favor of the sample of origin's covariate, we added the default omit_original_sample=True to avoid incorporating the sample of origin's log prob into the calculation.

CC @rastogiruchir

### CODE REVIEW

File: src/scvi/external/mrvi/_module.py [hunk 1]
```diff
@@ -517,7 +520,9 @@ def generative(
                 10.0 * jax.nn.one_hot(label_index, self.n_labels) if self.n_labels >= 2 else 0.0
             )
             cats = dist.Categorical(logits=self.u_prior_logits + offset)
-            normal_dists = dist.Normal(self.u_prior_means, jnp.exp(self.u_prior_scales))
+            normal_dists = dist.Normal(self.u_prior_means, jnp.exp(self.u_prior_scales)).to_event(
```
Comment (PierreBoyeau) [inline src/scvi/external/mrvi/_module.py:523]: snippet to double-check it works in `loss` (it works):

```
        pu = generative_outputs["pu"]
        
        # approach 1: below is the reference implementation of the log_prob of the mixture distribution
        log_pu = generative_outputs["pu"].log_prob(inference_outputs["u"])
        
        # approach 2: manual computation of the log_prob of the mixture distribution
        mixing_distribution = pu._mixing_distribution
        component_distribution = pu._component_distribution
        pk = mixing_distribution.probs
        mus = component_distribution.base_dist.loc
        stds = component_distribution.base_dist.scale
        
        pu_k = dist.Normal(mus, stds)
        log_pu_k = pu_k.log_prob(inference_outputs["u"][:, None, :]).sum(-1)
        log_pk = jnp.log(pk)
        log_puk = log_pk + log_pu_k
        # above are joint probas
        # shape (n_cells, n_mixtures)
        
        log_prob_u = jax.scipy.special.logsumexp(log_puk, axis=1)  # shape (n_cells,)
        # (log_prob_u == log_pu) ?
```

File: src/scvi/external/mrvi/_model.py [hunk 2]
```diff
@@ -789,11 +789,11 @@ def get_aggregated_posterior(
             qu_locs.append(outputs["qu"].loc)
             qu_scales.append(outputs["qu"].scale)
 
-        qu_loc = jnp.concatenate(qu_locs, axis=0).T
-        qu_scale = jnp.concatenate(qu_scales, axis=0).T
+        qu_loc = jnp.concatenate(qu_locs, axis=0)
```
Comment (PierreBoyeau) [inline src/scvi/external/mrvi/_model.py:792]: Add inline comments to clarify shapes

**Review by PierreBoyeau — APPROVED** [review]
LGTM

File: src/scvi/external/mrvi/_model.py [hunk 3]
```diff
@@ -916,8 +924,21 @@ def differential_abundance(
...
-                sel_log_probs = log_probs_arr.log_probs.loc[{"sample": cov_samples}]
-                val_log_probs = logsumexp(sel_log_probs, axis=1) - np.log(sel_log_probs.shape[1])
+                sel_log_probs = log_probs_arr.log_probs.loc[{"sample": cov_samples}].values
+                if omit_original_sample:
+                    sample_cov_one_hot = np.zeros((adata.n_obs, len(cov_samples)))
+                    for i, sample in enumerate(cov_samples):
+                        sample_cov_one_hot[adata.obs[self.sample_key] == sample, i] = 1
+                    sel_log_probs_no_original = np.where(
+                        sample_cov_one_hot, -np.inf, sel_log_probs
+                    )
```
Comment (PierreBoyeau) [inline src/scvi/external/mrvi/_model.py:934]:
```suggestion
                    )  # virtually discards samples-of-origin from aggregate posteriors
```

File: src/scvi/external/mrvi/_model.py [hunk 4]
```diff
@@ -930,10 +951,21 @@ def differential_abundance(
...
+                    if omit_original_sample:
+                        rest_samples_one_hot = np.zeros((adata.n_obs, len(rest_samples)))
+                        for i, sample in enumerate(rest_samples):
+                            rest_samples_one_hot[adata.obs[self.sample_key] == sample, i] = 1
+                        rest_log_probs_no_original = np.where(
+                            rest_samples_one_hot, -np.inf, rest_log_probs
+                        )
+                        rest_val_log_probs = logsumexp(
+                            rest_log_probs_no_original, axis=1
+                        ) - np.log((1 - rest_samples_one_hot).sum(axis=1))
```
Comment (PierreBoyeau) [inline src/scvi/external/mrvi/_model.py:964]: If this is exact duplicate from above, should we create a private function to mask these out?

**Review by PierreBoyeau — APPROVED** [review]
Looks good to me, see comments!

**Comment by PierreBoyeau** [comment]
Quick observations:

- These changes do not seem to affect the performed analyses conducted in the paper. Here are the log ratios (COVID vs  healthy) from the **complete** Haniffa dataset, showing a similar behavior as in the MrVI manuscript.
image
- That being said, these log ratios seem slightly different when computed on the drastically subsampled Haniffa dataset. If similar issues are reported on small datasets, we might want to explore options to smooth log ratios then.
